### PR TITLE
Develop ck

### PR DESCRIPTION
**Pull request by cknoll**

### CODE REVIEW

File: Code/gradcamheatmap.py [hunk 1]
```diff
@@ -1,62 +1,57 @@
...
+# our own modules
 
 # Import Gradcam - Check the gradcam.py for the code.
+from ATSDS import ATSDS
 from gradcam import get_gradcam
+from model import get_model
+import utils
 
-parser = argparse.ArgumentParser()
-parser.add_argument("--model_full_name", type=str, required=True, help="Full model name (e.g., simple_cnn_1_1)")
```
Comment (cknoll) [inline Code/gradcamheatmap.py:32]: This is now handled by `utils.get_default_arg_parser()`  (see below) because we use the same functionality also in `inference.py` and I try to adhere to the DRY-principle: don't repeat yourself.

File: Code/gradcamheatmap.py [hunk 2]
```diff
@@ -1,62 +1,57 @@
...
-model_name = "_".join(args.model_full_name.split("_")[:-2])
-model_cpt = args.model_full_name + ".tar"
 
-# DATASET_PATH = "data"
-DATASET_PATH = "/data/horse/ws/knoll-traffic_sign_reproduction"
-dataset_type = "atsds_large"
-dataset_split = "test"
+try:
+    # some optional debugging helpers
+    from ipydex import IPS, activate_ips_on_exception
```
Comment (cknoll) [inline Code/gradcamheatmap.py:38]: `ipydex` is a personal debugging and exploring library. I really find it useful but it is optional for this module.

You can find more information at https://github.com/cknoll/ipydex

File: Code/gradcamheatmap.py [hunk 3]
```diff
@@ -65,72 +60,6 @@ def load_model(model,optimizer,scheduler,filepath):
     optimizer.load_state_dict(cpt['optimizer'])
     scheduler.load_state_dict(cpt['scheduler'])
     return cpt['epoch'], cpt['trainstats']
-    
-RANDOM_SEED = 1414
-
-torch.manual_seed(RANDOM_SEED)
```
Comment (cknoll) [inline Code/gradcamheatmap.py:71]: all this global code is move to the `main` function

File: Code/gradcamheatmap.py [hunk 4]
```diff
@@ -157,7 +86,7 @@ def get_percentage_of_image(image,mask,percentage):
     masked_image = np.zeros_like(image)
     n = mask.size
     sortedvalues = np.sort(mask.flatten('K'))[::-1]
-    
+
```
Comment (cknoll) [inline Code/gradcamheatmap.py:89]: such edits (deleting trailing space characters on an otherwise empty line)  are caused by my editor configuration (deletes trailing spaces on every save). I would recommend to configure yours also like this. It is more professional to avoid pointless spaces.

File: Code/gradcamheatmap.py [hunk 5]
```diff
@@ -172,59 +101,161 @@ def get_input_tensors(image):
...
-elif model_name == "convnext":
-    GRADCAM_TARGET_LAYER = model.features[-1][-1].block[0]
-elif model_name == "vgg16":
-    GRADCAM_TARGET_LAYER = model.features[-3]
-    
-# GRADCAM_TARGET_LAYER = model.features[-1][-1].block[0] # convnext
-#GRADCAM_TARGET_LAYER = model.features[-3] # VGG
-#GRADCAM_TARGET_LAYER = model.layer4[-1].conv3 # resnet50
-print(GRADCAM_TARGET_LAYER)
+def main():
```
Comment (cknoll) [inline Code/gradcamheatmap.py:104]: This `main`-function now contains all the code from the global scope.

**Note:** to be a good function in terms of code-quality and maintainability it should be refactored into several smaller functions. My rule of thumb is:

- a function bigger than ca. 30 lines should be considered for modularization
- with > 50 lines or more then 3 levels of additional indentation even more.
- if breaking up into smaller functions is difficult because too many arguments and/or return-values would be involved, then I usually create a class and store the variables as instance attributes and then create methods instead of functions.
- main advantage: separate functions or methods can be tested and documented  separately. And they have their (hopefully descriptive name) which makes understanding the whole algorithm easier.

That said: I do not think it is necessary to refactor the `main` function now. It is just a general comment for this project and maybe future projects.

File: Code/gradcamheatmap.py [hunk 6]
```diff
@@ -172,59 +101,161 @@ def get_input_tensors(image):
...
+    dataset_type = "atsds_large"
+    dataset_split = "test"
+
+    CHECKPOINT_PATH = "model/"
+
 
-# Make sure the Output_path exists.
-# Define what percentage cutouts you want. 
-# If you are only interested in the masks you can just make this an empty list []
+    #################
```
Comment (cknoll) [inline Code/gradcamheatmap.py:129]: I used this marker to separate the different parts of the former global-scope code to allow for easier comparison

File: Code/gradcamheatmap.py [hunk 7]
```diff
@@ -172,59 +101,161 @@ def get_input_tensors(image):
...
+
+    # This creates the needed folders inside. As mentioned above the Folder defined in output_path has to already exist
+
+
+    if not os.path.isdir(output_path):
+        os.makedirs(output_path)
+
+    for cat in CATEGORIES:
+        for outputs in output_types:
+            os.makedirs(os.path.join(output_path, cat, outputs), exist_ok=True)
```
Comment (cknoll) [inline Code/gradcamheatmap.py:238]: Please note the `exists_ok=True` keyword argument. It replaces the former `if not os.path.isdir(output_path + cat + "/" + outputs):` statement.

File: Code/gradcamheatmap.py [hunk 8]
```diff
@@ -172,59 +101,161 @@ def get_input_tensors(image):
...
+    model_name = "_".join(args.model_full_name.split("_")[:-2])
+    model_cpt = args.model_full_name + ".tar"
+
+    # DATASET_PATH = "data"
+    if args.data_base_path is None:
+        # if argument is not passed: use hardcoded default
+        DATASET_PATH = "/data/horse/ws/knoll-traffic_sign_reproduction"
+    else:
+        DATASET_PATH = args.data_base_path
+
```
Comment (cknoll) [inline Code/gradcamheatmap.py:122]: The above `if` statement make the `--data_base_path` argument for the command line optional. However, I have not tested this.

File: Code/gradcamheatmap.py [hunk 9]
```diff
@@ -172,59 +101,161 @@ def get_input_tensors(image):
...
+    test_loss = trainstats[1]
+    detailed_stats = trainstats[2]
+
+    plt.plot(np.divide(train_loss,450), label = "train")
+    plt.plot(test_loss, label = "test")
+    plt.legend()
+
+    # IMAGES_PATH = 'data/' + dataset_type + '/' + dataset_split + '/'
+    # IMAGES_PATH = '/data/horse/ws/knoll-traffic_sign_reproduction/' + dataset_type + '/' + dataset_split + '/'
+    IMAGES_PATH = os.path.join(DATASET_PATH, dataset_type, dataset_split)
```
Comment (cknoll) [inline Code/gradcamheatmap.py:180]: Note, that I used `os.path.join(...)` instead of string concatenation via `+`. As far as I know it is the recommended way to assemble paths, because the result is adapted to the execution platform (e.g. win or unix). Also the reader immediately notices that a path is created.

File: Code/inference.py [hunk 10]
```diff
@@ -209,13 +209,7 @@ def main(model_full_name, data_base_path=None, model_cp_base_path=None, mode="co
...
 
-    parser = argparse.ArgumentParser()
-    # for every argument we also have a short form
-    parser.add_argument(
-        "--model_full_name", "-n", type=str, required=True, help="Full model name (e.g., simple_cnn_1_1)"
-    )
-    parser.add_argument("--model_cp_base_path", "-cp", type=str, help="directory of model checkpoints", default=None)
-    parser.add_argument("--data_base_path", "-d", type=str, help="data path", default=None)
+    parser = utils.get_default_arg_parser()
     parser.add_argument(
```
Comment (cknoll) [inline Code/inference.py:213]: We use our default parser from `utils.py` (see below). Reason: DRY principle.

Note: We can still add more arguments to this `parser` object.

File: README.md [hunk 11]
```diff
@@ -8,8 +8,19 @@ This repo contains code based on the master thesis "Evaluation of XAI-Algorithms
...
+
 Local usage:
 - `python inference.py py3 --model_full_name simple_cnn_1_1 -d ./`
 
 Usage with mounted group drive and json-mode (see `--help`) :
 - `python inference.py --model_full_name simple_cnn_1_1 -cp /home/ck/mnt/XAI-DIA-gl/Julian/Dataset_Masterarbeit/model_checkpoints -d /home/ck/mnt/XAI-DIA-gl/Julian/Dataset_Masterarbeit/atsds_large -m json`
+
+
+#### Creating gradcam heatmaps
+
```
Comment (cknoll) [inline README.md:21]: I find it useful to document some typical command in the README.md. For convenience I use the paths specific for my system. You can of course also add commands with your paths.
